### PR TITLE
Implement replacement policies for vmemcache

### PR DESCRIPTION
**Pull request by ldorau**
This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/6)

### CODE REVIEW

**Comment by ldorau** [comment]
This PR contains PR https://github.com/pmem/vmemcache/pull/4 yet.

File: src/vmemcache_repl.c [hunk 1]
```diff
@@ -174,6 +174,19 @@ repl_p_none_evict(struct repl_p_head *head, struct repl_p_entry *entry)
 static int
 repl_p_lru_new(struct repl_p_head **head)
 {
+	struct repl_p_head *h = Zalloc(sizeof(struct repl_p_head));
+	if (h == NULL)
+		return -1;
+
+	util_mutex_init(&h->lock);
+	util_mutex_lock(&h->lock);
```
Comment (pbalcer) [inline src/vmemcache_repl.c]: you don't need to grab locks at new and delete ;)

File: src/vmemcache_repl.c [hunk 2]
```diff
@@ -183,6 +196,18 @@ repl_p_lru_new(struct repl_p_head **head)
...
+	util_mutex_lock(&head->lock);
+
+	while (!STAILQ_EMPTY(&head->first)) {
+		struct repl_p_entry *entry = STAILQ_FIRST(&head->first);
+		STAILQ_REMOVE_HEAD(&head->first, node);
+		Free(entry);
+	}
+
+	util_mutex_unlock(&head->lock);
+	util_mutex_destroy(&head->lock);
```
Comment (marcinslusarz) [inline src/vmemcache_repl.c]: If you need to take a lock before destroy, something is seriously wrong.

Comment (ldorau) [reply src/vmemcache_repl.c]: .

File: src/vmemcache_repl.c [hunk 3]
```diff
@@ -200,6 +228,16 @@ repl_p_lru_insert(struct repl_p_head *head, void *element)
 static int
 repl_p_lru_use(struct repl_p_head *head, struct repl_p_entry *entry)
 {
+	if (entry == NULL)
```
Comment (pbalcer) [inline src/vmemcache_repl.c]: `ASSERTne(entry, NULL);`
You could also add:
`ASSERT(entry exists in head);`
And then make this function return void.

File: src/vmemcache_repl.c [hunk 4]
```diff
@@ -209,5 +247,22 @@ repl_p_lru_use(struct repl_p_head *head, struct repl_p_entry *entry)
 static void *
 repl_p_lru_evict(struct repl_p_head *head, struct repl_p_entry *entry)
 {
-	return NULL;
+	void *data = NULL;
+
+	util_mutex_lock(&head->lock);
+
+	if (!entry) {
+		entry = STAILQ_FIRST(&head->first);
```
Comment (pbalcer) [inline src/vmemcache_repl.c]:
```
if (entry == NULL && (entry = STAILQ_FIRST(&head->first)) == NULL)
  goto exit_unlock;
```

Comment (ldorau) [reply src/vmemcache_repl.c]: Done

**Pull request merged**